From aa48e52daf219c224648528053ec41c358930f16 Mon Sep 17 00:00:00 2001
From: Florian Festi <ffesti@redhat.com>
Date: Wed, 21 Feb 2024 08:25:27 +0100
Subject: [PATCH 1/5] Add --list and --delete to rpmkeys

This is a  bit of a hack as it manipulates the parsed cli parameters to
to the "right thing" and then calls rpmcliQuery and rpmErase.

(cherry picked from commit 1dc7e76fa51bca54f0eb75660ab6e68216289eb6)
---
 docs/man/rpmkeys.8.md | 18 +++++++++++++++---
 tools/rpmkeys.c       | 40 ++++++++++++++++++++++++++++++++++------
 2 files changed, 49 insertions(+), 9 deletions(-)

diff --git a/docs/man/rpmkeys.8.md b/docs/man/rpmkeys.8.md
index 9ea0a2079..cbc619001 100644
--- a/docs/man/rpmkeys.8.md
+++ b/docs/man/rpmkeys.8.md
@@ -12,15 +12,19 @@ rpmkeys - RPM Keyring
 SYNOPSIS
 ========
 
-**rpmkeys** {**\--import\|\--checksig**}
+**rpmkeys** {**\--list\|\--import\|\--delete\|\--checksig**}
 
 DESCRIPTION
 ===========
 
 The general forms of rpm digital signature commands are
 
+**rpmkeys** **\--list** \[*KEYHASH \...*\]
+
 **rpmkeys** **\--import** *PUBKEY \...*
 
+**rpmkeys** **\--delete** *KEYHASH \...*
+
 **rpmkeys** {**-K\|\--checksig**} *PACKAGE\_FILE \...*
 
 The **\--checksig** option checks all the digests and signatures
@@ -37,13 +41,21 @@ example, all currently imported public keys can be displayed by:
 
 **rpm -qa gpg-pubkey\***
 
-Details about a specific public key, when imported, can be displayed by
+A more convenient way to display them is
+
+**rpmkeys** **\--list**
+
+More details about a specific public key, when imported, can be displayed by
 querying. Here\'s information about the Red Hat GPG/DSA key:
 
 **rpm -qi gpg-pubkey-db42a60e**
 
 Finally, public keys can be erased after importing just like packages.
-Here\'s how to remove the Red Hat GPG/DSA key
+Here\'s how to remove the Red Hat GPG/DSA key:
+
+**rpmkeys** **\--delete db42a60e**
+
+Or alternatively:
 
 **rpm -e gpg-pubkey-db42a60e**
 
diff --git a/tools/rpmkeys.c b/tools/rpmkeys.c
index afaffe501..adb65735c 100644
--- a/tools/rpmkeys.c
+++ b/tools/rpmkeys.c
@@ -2,6 +2,7 @@
 
 #include <popt.h>
 #include <rpm/rpmcli.h>
+#include <rpm/rpmstring.h>
 #include "cliutils.h"
 #include "debug.h"
 
@@ -22,12 +23,10 @@ static struct poptOption keyOptsTable[] = {
 	N_("import an armored public key"), NULL },
     { "test", '\0', POPT_ARG_NONE, &test, 0,
 	N_("don't import, but tell if it would work or not"), NULL },
-#if 0
-    { "delete-key", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
+    { "delete", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
+	N_("delete keys from RPM keyring"), NULL },
+    { "list", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_LISTKEY,
 	N_("list keys from RPM keyring"), NULL },
-    { "list-keys", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_LISTKEY,
-	N_("list keys from RPM keyring"), NULL },
-#endif
     POPT_TABLEEND
 };
 
@@ -42,6 +41,21 @@ static struct poptOption optionsTable[] = {
     POPT_TABLEEND
 };
 
+static ARGV_t gpgkeyargs(ARGV_const_t args) {
+    ARGV_t gpgargs = NULL;
+    for (char * const * arg = args; *arg; arg++) {
+	if (strncmp(*arg, "gpg-pubkey-", 11)) {
+	    char * gpgarg = NULL;
+	    rstrscat(&gpgarg, "gpg-pubkey-", *arg, NULL);
+	    argvAdd(&gpgargs, gpgarg);
+	    free(gpgarg);
+	} else {
+	    argvAdd(&gpgargs, *arg);
+	}
+    }
+    return gpgargs;
+}
+
 int main(int argc, char *argv[])
 {
     int ec = EXIT_FAILURE;
@@ -73,9 +87,23 @@ int main(int argc, char *argv[])
 	    rpmtsSetFlags(ts, (rpmtsFlags(ts)|RPMTRANS_FLAG_TEST));
 	ec = rpmcliImportPubkeys(ts, args);
 	break;
-    /* XXX TODO: actually implement these... */
     case MODE_DELKEY:
+	struct rpmInstallArguments_s * ia = &rpmIArgs;
+	ARGV_t gpgargs = gpgkeyargs(args);
+	ec = rpmErase(ts, ia, gpgargs);
+	argvFree(gpgargs);
+	break;
     case MODE_LISTKEY:
+	ARGV_t query = NULL;
+	if (args != NULL) {
+	    query = gpgkeyargs(args);
+	} else {
+	    argvAdd(&query, "gpg-pubkey");
+	}
+	QVA_t qva = &rpmQVKArgs;
+	rstrcat(&qva->qva_queryFormat, "%{version}-%{release}: %{summary}\n");
+	ec = rpmcliQuery(ts, &rpmQVKArgs, (ARGV_const_t) query);
+	query = argvFree(query);
 	break;
     default:
 	argerror(_("only one major mode may be specified"));
-- 
2.50.1


From a04cbdf459b75e326ad9a3e78cabd2980c5c410e Mon Sep 17 00:00:00 2001
From: Panu Matilainen <pmatilai@redhat.com>
Date: Tue, 15 Oct 2024 09:28:36 +0300
Subject: [PATCH 2/5] Having no keys imported is not an error

...any more than "ls" in an empty directory is.

Resolves: 3556
Co-authored-by: Florian Festi <ffesti@redhat.com>

(backported from commit d666883624c5f2905c0bc70112895c33605ca264)
---
 tools/rpmkeys.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/tools/rpmkeys.c b/tools/rpmkeys.c
index adb65735c..d2c2644e2 100644
--- a/tools/rpmkeys.c
+++ b/tools/rpmkeys.c
@@ -95,12 +95,13 @@ int main(int argc, char *argv[])
 	break;
     case MODE_LISTKEY:
 	ARGV_t query = NULL;
+	QVA_t qva = &rpmQVKArgs;
 	if (args != NULL) {
 	    query = gpgkeyargs(args);
 	} else {
+	    qva->qva_source |= RPMQV_ALL;
 	    argvAdd(&query, "gpg-pubkey");
 	}
-	QVA_t qva = &rpmQVKArgs;
 	rstrcat(&qva->qva_queryFormat, "%{version}-%{release}: %{summary}\n");
 	ec = rpmcliQuery(ts, &rpmQVKArgs, (ARGV_const_t) query);
 	query = argvFree(query);
-- 
2.50.1


From cd0cf8fbdde64656b851ad6056f0c40e18c90a46 Mon Sep 17 00:00:00 2001
From: Florian Festi <ffesti@redhat.com>
Date: Fri, 15 Nov 2024 13:15:58 +0100
Subject: [PATCH 3/5] Add short CLI commands to rpmkeys

Switch to short versions in the test suite at a few places.

Resolves: #3435
(backported from commit 35cfb7de9842dbec35adcb9ad9a632df49f35102)
---
 docs/man/rpmkeys.8.md | 6 +++---
 tools/rpmkeys.c       | 8 ++++----
 2 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/docs/man/rpmkeys.8.md b/docs/man/rpmkeys.8.md
index cbc619001..77bce49a9 100644
--- a/docs/man/rpmkeys.8.md
+++ b/docs/man/rpmkeys.8.md
@@ -19,11 +19,11 @@ DESCRIPTION
 
 The general forms of rpm digital signature commands are
 
-**rpmkeys** **\--list** \[*KEYHASH \...*\]
+**rpmkeys** {**-l\|\--list**} \[*KEYHASH \...*\]
 
-**rpmkeys** **\--import** *PUBKEY \...*
+**rpmkeys** {**-i\|\--import**} *PUBKEY \...*
 
-**rpmkeys** **\--delete** *KEYHASH \...*
+**rpmkeys** {**-d\|\--delete**} *KEYHASH \...*
 
 **rpmkeys** {**-K\|\--checksig**} *PACKAGE\_FILE \...*
 
diff --git a/tools/rpmkeys.c b/tools/rpmkeys.c
index d2c2644e2..79bb3e236 100644
--- a/tools/rpmkeys.c
+++ b/tools/rpmkeys.c
@@ -19,13 +19,13 @@ static int test = 0;
 static struct poptOption keyOptsTable[] = {
     { "checksig", 'K', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_CHECKSIG,
 	N_("verify package signature(s)"), NULL },
-    { "import", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_IMPORTKEY,
+    { "import", 'i', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_IMPORTKEY,
 	N_("import an armored public key"), NULL },
-    { "test", '\0', POPT_ARG_NONE, &test, 0,
+    { "test", 't', POPT_ARG_NONE, &test, 0,
 	N_("don't import, but tell if it would work or not"), NULL },
-    { "delete", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
+    { "delete", 'd', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
 	N_("delete keys from RPM keyring"), NULL },
-    { "list", '\0', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_LISTKEY,
+    { "list", 'l', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_LISTKEY,
 	N_("list keys from RPM keyring"), NULL },
     POPT_TABLEEND
 };
-- 
2.50.1


From 8b344dfe14eda275e85f3acece93f98f5cab1e37 Mon Sep 17 00:00:00 2001
From: Florian Festi <ffesti@redhat.com>
Date: Fri, 15 Nov 2024 14:03:36 +0100
Subject: [PATCH 4/5] Rename rpmkeys --delete to rpmkeys --erase

Keep in line with the command names in rpm itself.
Keep --delete functional for compatibility reasons but remove it from
the docs.

Related: #3435
(backported from commit abb6ab1c43a946eb36a36227e39918580390a8e4)
---
 docs/man/rpmkeys.8.md | 4 ++--
 tools/rpmkeys.c       | 7 +++++--
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/docs/man/rpmkeys.8.md b/docs/man/rpmkeys.8.md
index 77bce49a9..393845e62 100644
--- a/docs/man/rpmkeys.8.md
+++ b/docs/man/rpmkeys.8.md
@@ -12,7 +12,7 @@ rpmkeys - RPM Keyring
 SYNOPSIS
 ========
 
-**rpmkeys** {**\--list\|\--import\|\--delete\|\--checksig**}
+**rpmkeys** {**\--list\|\--import\|\--erase\|\--checksig**}
 
 DESCRIPTION
 ===========
@@ -23,7 +23,7 @@ The general forms of rpm digital signature commands are
 
 **rpmkeys** {**-i\|\--import**} *PUBKEY \...*
 
-**rpmkeys** {**-d\|\--delete**} *KEYHASH \...*
+**rpmkeys** {**-e\|\--erase**} *KEYHASH \...*
 
 **rpmkeys** {**-K\|\--checksig**} *PACKAGE\_FILE \...*
 
diff --git a/tools/rpmkeys.c b/tools/rpmkeys.c
index 79bb3e236..77eaa703f 100644
--- a/tools/rpmkeys.c
+++ b/tools/rpmkeys.c
@@ -23,8 +23,11 @@ static struct poptOption keyOptsTable[] = {
 	N_("import an armored public key"), NULL },
     { "test", 't', POPT_ARG_NONE, &test, 0,
 	N_("don't import, but tell if it would work or not"), NULL },
-    { "delete", 'd', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
-	N_("delete keys from RPM keyring"), NULL },
+    { "delete", 'd', (POPT_ARG_VAL|POPT_ARGFLAG_OR|POPT_ARGFLAG_DOC_HIDDEN),
+        &mode, MODE_DELKEY,
+	N_("Erase keys from RPM keyring"), NULL },
+    { "erase", 'e', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
+	N_("Erase keys from RPM keyring"), NULL },
     { "list", 'l', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_LISTKEY,
 	N_("list keys from RPM keyring"), NULL },
     POPT_TABLEEND
-- 
2.50.1


From 2cd7d8d91d3c8bb7bf800bc7982ee2f5926d5c15 Mon Sep 17 00:00:00 2001
From: Florian Festi <ffesti@redhat.com>
Date: Fri, 15 Nov 2024 19:22:07 +0100
Subject: [PATCH 5/5] Don't delete delete

(backported from commit 4c99a010b23fd3832025d9cd2a5481a81f5c3353)
---
 docs/man/rpmkeys.8.md | 4 ++--
 tools/rpmkeys.c       | 3 +--
 2 files changed, 3 insertions(+), 4 deletions(-)

diff --git a/docs/man/rpmkeys.8.md b/docs/man/rpmkeys.8.md
index 393845e62..afa470b6f 100644
--- a/docs/man/rpmkeys.8.md
+++ b/docs/man/rpmkeys.8.md
@@ -12,7 +12,7 @@ rpmkeys - RPM Keyring
 SYNOPSIS
 ========
 
-**rpmkeys** {**\--list\|\--import\|\--erase\|\--checksig**}
+**rpmkeys** {**\--list\|\--import\|\--erase\|\--delete\|\--checksig**}
 
 DESCRIPTION
 ===========
@@ -23,7 +23,7 @@ The general forms of rpm digital signature commands are
 
 **rpmkeys** {**-i\|\--import**} *PUBKEY \...*
 
-**rpmkeys** {**-e\|\--erase**} *KEYHASH \...*
+**rpmkeys** {**-e\|\--erase\|-d\|\--delete**} *KEYHASH \...*
 
 **rpmkeys** {**-K\|\--checksig**} *PACKAGE\_FILE \...*
 
diff --git a/tools/rpmkeys.c b/tools/rpmkeys.c
index 77eaa703f..f425a8648 100644
--- a/tools/rpmkeys.c
+++ b/tools/rpmkeys.c
@@ -23,8 +23,7 @@ static struct poptOption keyOptsTable[] = {
 	N_("import an armored public key"), NULL },
     { "test", 't', POPT_ARG_NONE, &test, 0,
 	N_("don't import, but tell if it would work or not"), NULL },
-    { "delete", 'd', (POPT_ARG_VAL|POPT_ARGFLAG_OR|POPT_ARGFLAG_DOC_HIDDEN),
-        &mode, MODE_DELKEY,
+    { "delete", 'd', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
 	N_("Erase keys from RPM keyring"), NULL },
     { "erase", 'e', (POPT_ARG_VAL|POPT_ARGFLAG_OR), &mode, MODE_DELKEY,
 	N_("Erase keys from RPM keyring"), NULL },
-- 
2.50.1

diff -up rpm-4.19.1.1/docs/man/rpmkeys.8.orig rpm-4.19.1.1/docs/man/rpmkeys.8
--- rpm-4.19.1.1/docs/man/rpmkeys.8.orig	2025-07-24 16:37:03.969960189 +0200
+++ rpm-4.19.1.1/docs/man/rpmkeys.8	2025-07-24 16:37:28.946868811 +0200
@@ -1,79 +1,73 @@
-.\" Automatically generated by Pandoc 3.1.3
+.\" Automatically generated by Pandoc 3.1.11.1
 .\"
-.\" Define V font for inline verbatim, using C font in formats
-.\" that render this, and otherwise B font.
-.ie "\f[CB]x\f[]"x" \{\
-. ftr V B
-. ftr VI BI
-. ftr VB B
-. ftr VBI BI
-.\}
-.el \{\
-. ftr V CR
-. ftr VI CI
-. ftr VB CB
-. ftr VBI CBI
-.\}
 .TH "RPMKEYS" "8" "29 October 2010" "" ""
-.hy
 .SH NAME
-.PP
-rpmkeys - RPM Keyring
+rpmkeys \- RPM Keyring
 .SH SYNOPSIS
-.PP
-\f[B]rpmkeys\f[R] {\f[B]--import|--checksig\f[R]}
+\f[B]rpmkeys\f[R]
+{\f[B]\-\-list|\-\-import|\-\-erase|\-\-delete|\-\-checksig\f[R]}
 .SH DESCRIPTION
-.PP
 The general forms of rpm digital signature commands are
 .PP
-\f[B]rpmkeys\f[R] \f[B]--import\f[R] \f[I]PUBKEY ...\f[R]
+\f[B]rpmkeys\f[R] {\f[B]\-l|\-\-list\f[R]} [\f[I]KEYHASH ...\f[R]]
+.PP
+\f[B]rpmkeys\f[R] {\f[B]\-i|\-\-import\f[R]} \f[I]PUBKEY ...\f[R]
+.PP
+\f[B]rpmkeys\f[R] {\f[B]\-e|\-\-erase|\-d|\-\-delete\f[R]} \f[I]KEYHASH
+\&...\f[R]
 .PP
-\f[B]rpmkeys\f[R] {\f[B]-K|--checksig\f[R]} \f[I]PACKAGE_FILE ...\f[R]
+\f[B]rpmkeys\f[R] {\f[B]\-K|\-\-checksig\f[R]} \f[I]PACKAGE_FILE
+\&...\f[R]
 .PP
-The \f[B]--checksig\f[R] option checks all the digests and signatures
+The \f[B]\-\-checksig\f[R] option checks all the digests and signatures
 contained in \f[I]PACKAGE_FILE\f[R] to ensure the integrity and origin
 of the package.
 Note that signatures are now verified whenever a package is read, and
-\f[B]--checksig\f[R] is useful to verify all of the digests and
+\f[B]\-\-checksig\f[R] is useful to verify all of the digests and
 signatures associated with a package.
 .PP
 Digital signatures cannot be verified without a public key.
 An ASCII armored public key can be added to the \f[B]rpm\f[R] database
-using \f[B]--import\f[R].
+using \f[B]\-\-import\f[R].
 An imported public key is carried in a header, and key ring management
 is performed exactly like package management.
 For example, all currently imported public keys can be displayed by:
 .PP
-\f[B]rpm -qa gpg-pubkey*\f[R]
+\f[B]rpm \-qa gpg\-pubkey*\f[R]
 .PP
-Details about a specific public key, when imported, can be displayed by
-querying.
+A more convenient way to display them is
+.PP
+\f[B]rpmkeys\f[R] \f[B]\-\-list\f[R]
+.PP
+More details about a specific public key, when imported, can be
+displayed by querying.
 Here\[aq]s information about the Red Hat GPG/DSA key:
 .PP
-\f[B]rpm -qi gpg-pubkey-db42a60e\f[R]
+\f[B]rpm \-qi gpg\-pubkey\-db42a60e\f[R]
 .PP
 Finally, public keys can be erased after importing just like packages.
-Here\[aq]s how to remove the Red Hat GPG/DSA key
+Here\[aq]s how to remove the Red Hat GPG/DSA key:
 .PP
-\f[B]rpm -e gpg-pubkey-db42a60e\f[R]
-.SH SEE ALSO
+\f[B]rpmkeys\f[R] \f[B]\-\-delete db42a60e\f[R]
 .PP
+Or alternatively:
+.PP
+\f[B]rpm \-e gpg\-pubkey\-db42a60e\f[R]
+.SH SEE ALSO
 \f[B]popt\f[R](3), \f[B]rpm\f[R](8), \f[B]rpmdb\f[R](8),
 \f[B]rpmsign\f[R](8), \f[B]rpm2cpio\f[R](8), \f[B]rpmbuild\f[R](8),
 \f[B]rpmspec\f[R](8)
 .PP
-\f[B]rpmkeys --help\f[R] - as rpm supports customizing the options via
-popt aliases it\[aq]s impossible to guarantee that what\[aq]s described
-in the manual matches what\[aq]s available.
+\f[B]rpmkeys \-\-help\f[R] \- as rpm supports customizing the options
+via popt aliases it\[aq]s impossible to guarantee that what\[aq]s
+described in the manual matches what\[aq]s available.
 .PP
 \f[B]http://www.rpm.org/ <URL:http://www.rpm.org/>\f[R]
 .SH AUTHORS
 .IP
-.nf
-\f[C]
+.EX
 Marc Ewing <marc\[at]redhat.com>
 Jeff Johnson <jbj\[at]redhat.com>
 Erik Troan <ewt\[at]redhat.com>
 Panu Matilainen <pmatilai\[at]redhat.com>
-\f[R]
-.fi
+.EE
